chatgpt-shell-google.el

```diff
@@ -82,6 +83,110 @@ VALIDATE-COMMAND, and GROUNDING-SEARCH handler."
+                                         :grounding-search model-supports-grounding
+                                         :path model-urlpath
+                                         :token-width 4
+                                         :context-window model-cwindow)))))
+
+(cl-defun chatgpt-shell-google-load-models (&key override)
+  "Query Google for the list of Gemini LLM models available (see
+https://ai.google.dev/gemini-api/docs/models/gemini) and add them to
+`chatgpt-shell-models' unless a model with the same name is already
+present. By default, replace the existing Google models in
+`chatgpt-shell-models' with the newly retrieved models. When OVERRIDE is
+non-nil (interactively with a prefix argument), replace all the Google
+models with those retrieved."
+  (interactive (list :override current-prefix-arg))
+  (let* ((goog-predicate (lambda (model)
+                           (string= (map-elt model :provider) "Google")))
+         (goog-index (or (cl-position-if goog-predicate chatgpt-shell-models)
+                         (length chatgpt-shell-models))))
+    (setq chatgpt-shell-models (and (not override)
+                                    (cl-remove-if goog-predicate chatgpt-shell-models)))
+    (let* ((existing-gemini-models (mapcar (lambda (model)
+                                             (map-elt model :version))
+                                           (cl-remove-if-not goog-predicate
+                                                             chatgpt-shell-models)))
+           (new-gemini-models
+            (mapcar #'chatgpt-shell--google-convert-model (chatgpt-shell--google-get-generative-models))))
+      (setq chatgpt-shell-models
+            (append (seq-take chatgpt-shell-models goog-index)
+                    new-gemini-models
+                    (seq-drop chatgpt-shell-models goog-index)))
+      (message "Added %d Gemini model(s); kept %d existing Gemini model(s)"
+               (length new-gemini-models)
+               (length existing-gemini-models)))))
+
+(defun chatgpt-shell-google-toggle-grounding ()
+  "Toggles the :grounding for the currently-selected model. Google's
+documentation states that All Gemini 1.5 and 2.0 models support
+grounding, some of the experimental or short-lived models do not.  If
+chatgpt-shell tries to use a model that does nto support grounding, the
+API returns an error. In that case, the user can toggle grounding on the
+model, using this function."
+  (interactive)
+  (let ((current-model (chatgpt-shell--resolved-model)))
+    (when (and current-model
+               (string= (map-elt current-model :provider) "Google"))
+      (let ((current-grounding-cons
+             (assq :grounding-search current-model)))
+        (when current-grounding
+          (setf (cdr current-grounding-cons) (not (cdr current-grounding-cons))))))))
+
+(defun chatgpt-shell-google-get-grounding-tool-keyword ()
```

Comment:
Consider passing the `model` to make this function stateless. The calling function already has the `model` handy.

Reply:
done